Fix apply preview and exceptions when switching loadouts

Al12rs opened the pull request:
Preview view was showing wrong data, now it should show all the changes that will happen to the last analyzed disk state if the loadout is applied. 

This means that if something is on disk but not in the loadout it will appear as a normal entry in the preview, since it will not change. The preview will only show how the game folder will be affected by Apply, not how the loadout will be affected by apply.

I added a different method to determine whether Synchronize should be needed, since the previous one was ignoring a number of cases that still required a synchronize.

Also fixed exceptions that would happen due to missing LastAppliedLoadout data after switching lodouts.

halgari commented on src/Abstractions/NexusMods.Abstractions.Loadouts.Synchronizers/ALoadoutSynchronizer.cs:
```diff
@@ -1471,7 +1502,7 @@ public async Task ResetToOriginalGameState(GameInstallation installation, string
 
         // Run the groupings
-        await RunActions(desiredState, installation);
+        await RunActions(desiredState, installation, clearLastSyncedLoadout);
```
I'd like some documentation and/or tests on why this change is required. It adds another dimetion to this code that I think may make it harder to debug in the long run. Why can't always clear or not clear the state? Also if we've reset the game state, wouldn't the last applied state no longer be valid?

Al12rs replied:
The problem is that this code is currently being used to switch loadouts, in which case we don't want to reset the state.

The best solution would probably be to have a dedicated optimized code path to switch loadouts, rather than resetting the game folder to default and then applying the new loadout, but I think that would be outside the scope of this.

In the case of a loadout switch we don't want to delete the `LastSyncedLoadoutTransaction` and `LastSyncedLoadout`, as they will still be needed to determine the loadout Synchronized state (should apply).

I think the current loadout switching tests don't catch this since the errors happen in the `ShouldApply` check, that is only performed if there is at least a subscription active on `StatusForGame()` observable. 

I'll try adding that to the tests.

halgari commented:
Why don't we want to reset the states? If we're swapping loadouts (a
relatively rare occurrence) we're doing exactly what this code says:
deactivate the old account and reactivate the new one. What I'm trying to
avoid here is introducing a difference between applying a loadout for the
first time and swapping a loadout. The fewer unique states we have the more
consistent testing should be.


On Mon, Feb 17, 2025 at 7:11 AM Al ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In
> src/Abstractions/NexusMods.Abstractions.Loadouts.Synchronizers/ALoadoutSynchronizer.cs
> <https://github.com/Nexus-Mods/NexusMods.App/pull/2637#discussion_r1958304951>
> :
>
> > @@ -1471,7 +1502,7 @@ public async Task ResetToOriginalGameState(GameInstallation installation, string
>          ProcessSyncTree(desiredState);
>
>          // Run the groupings
> -        await RunActions(desiredState, installation);
> +        await RunActions(desiredState, installation, clearLastSyncedLoadout);
>
> The problem is that this code is currently being used to switch loadouts,
> in which case we don't want to reset the state.
>
> The best solution would probably be to have a dedicated optimized code
> path to switch loadouts, rather than resetting the game folder to default
> and then applying the new loadout, but I think that would be outside the
> scope of this.
>
> In the case of a loadout switch we don't want to delete the
> LastSyncedLoadoutTransaction and LastSyncedLoadout, as they will still be
> needed to determine the loadout Synchronized state (should apply).
>
> I think the current loadout switching tests don't catch this since the
> errors happen in the ShouldApply check, that is only performed if there
> is at least a subscription active on StatusForGame() observable.
>
> I'll try adding that to the tests.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Nexus-Mods/NexusMods.App/pull/2637#discussion_r1958304951>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAE72HN2JUWORPJZZIS4N7T2QHUYFAVCNFSM6AAAAABXBZBWHKVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDMMRRGE4DGMJTGY>
> .
> You are receiving this because your review was requested.Message ID:
> ***@***.***>
>

Al12rs commented:
> Why don't we want to reset the states? If we're swapping loadouts (a relatively rare occurrence) we're doing exactly what this code says: deactivate the old account and reactivate the new one. What I'm trying to avoid here is introducing a difference between applying a loadout for the first time and swapping a loadout. The fewer unique states we have the more consistent testing should be.

The code that checks the loadout status (`ShouldSynchronize()` and `LoadoutToDiskDiff()`) needs to know what the previous `DiskState` was. Without that information the results become inaccurate.

After a game is managed, that code assumes we always have a loadout applied, up until we unmanage the game. 
The only case where resetting to default game state should happen is either as an intermediate step of switching loadouts, or when unmanaging the game.

There should be no need to clear the last `LastSyncedLoadout` and the `LastSyncedLoadoutTransaction` when switching loadouts. That is still information we need for the purposes of computing the synchronize status of loadouts and diffs.


I updated the loadout switching test to more accurately follow what the app does when switching loadouts and I added the subscription to the state of the loadouts to listen to any errors that happen there.

That test now catches the exceptions that this PR fixes.
I also found that the  `StubbedFileHasherService` was regenerating the game indexing data every call, which was causing issues.

Al12rs commented:
I changed the loadout sync state functions to fallback to the last indexed state if the `LastSynched` state is missing, which is going to be incorrect, but the incorrect data should be replaced by more accurate data once the loadout switch fully completes.

This way deactivating a loadout always does the same thing, and no exceptions are thrown during sync state checks
